### PR TITLE
Added Pangu, Fengu, SwinRNN models

### PR DESCRIPTION
**Pull request by ivanauyeung**
# Modulus Pull Request

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

## Dependencies

timm>=0.9.12

### CODE REVIEW

File: examples/weather/pangu_weather/train_pangu_era5.py [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+from modulus.utils import StaticCaptureTraining, StaticCaptureEvaluateNoGrad
+
+from modulus.launch.logging import LaunchLogger, PythonLogger, initialize_mlflow
+from modulus.launch.utils import load_checkpoint, save_checkpoint
+
+try:
+    from apex import optimizers
+except:
+    raise ImportError(
+        "FCN training requires apex package for optimizer."
```
Comment (dallasfoster) [inline examples/weather/pangu_weather/train_pangu_era5.py:39]: Change the import error here.

File: examples/weather/pangu_weather/train_pangu_era5.py [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+    DistributedManager.initialize()
+    dist = DistributedManager()
+
+    # Initialize loggers
+    # initialize_wandb(
+    #     project="Modulus-Launch-Dev",
+    #     entity="Modulus",
+    #     name="FourCastNet-Training",
+    #     group="FCN-DDP-Group",
+    # )
```
Comment (dallasfoster) [inline examples/weather/pangu_weather/train_pangu_era5.py:131]: Let's remove this commented code.

File: examples/weather/pangu_weather/train_pangu_era5.py [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+
+    no_channals_pangu = 4 + 5 * 13
+    datapipe = ERA5HDF5Datapipe(
+        data_dir="/data/train/",
+        stats_dir="/data/stats/",
+        channels=[i for i in range(no_channals_pangu)],
+        num_samples_per_year=cfg.num_samples_per_year_train,
+        batch_size=1,
+        patch_size=(1, 1),
+        num_workers=8,
```
Comment (dallasfoster) [inline examples/weather/pangu_weather/train_pangu_era5.py:151]: Can any of these be added to the config?

File: examples/weather/pangu_weather/train_pangu_era5.py [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+        device=dist.device,
+        process_rank=dist.rank,
+        world_size=dist.world_size,
+    )
+    logger.success(f"Loaded datapipe of size {len(datapipe)}")
+
+    mask_dir = "/data/constant_mask"
```
Comment (dallasfoster) [inline examples/weather/pangu_weather/train_pangu_era5.py:158]: Can this be a configurable?

File: examples/weather/pangu_weather/train_pangu_era5.py [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+            shuffle=False,
+        )
+        logger.success(f"Loaded validaton datapipe of size {len(validation_datapipe)}")
+
+    pangu_model = Pangu(
+        img_size=(721, 1440),
+        patch_size=(2, 4, 4),
+        embed_dim=192,
+        num_heads=(6, 12, 12, 6),
+        window_size=(2, 6, 12),
```
Comment (dallasfoster) [inline examples/weather/pangu_weather/train_pangu_era5.py:194]: Can any of this be configurable?

File: examples/weather/pangu_weather/train_pangu_lite_era5.py [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+    DistributedManager.initialize()
+    dist = DistributedManager()
+
+    # Initialize loggers
+    # initialize_wandb(
+    #     project="Modulus-Launch-Dev",
+    #     entity="Modulus",
+    #     name="FourCastNet-Training",
+    #     group="FCN-DDP-Group",
+    # )
```
Comment (dallasfoster) [inline examples/weather/pangu_weather/train_pangu_lite_era5.py:131]: Remove commented code here.

File: examples/weather/pangu_weather/train_pangu_lite_era5.py [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+
+    no_channals_pangu = 4 + 5 * 13
+    datapipe = ERA5HDF5Datapipe(
+        data_dir="/data/train/",
+        stats_dir="/data/stats/",
+        channels=[i for i in range(no_channals_pangu)],
+        num_samples_per_year=cfg.num_samples_per_year_train,
+        batch_size=1,
+        patch_size=(1, 1),
+        num_workers=8,
```
Comment (dallasfoster) [inline examples/weather/pangu_weather/train_pangu_lite_era5.py:151]: Same comment, can these be configurable.

File: examples/weather/pangu_weather/train_pangu_lite_era5.py [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+        device=dist.device,
+        process_rank=dist.rank,
+        world_size=dist.world_size,
+    )
+    logger.success(f"Loaded datapipe of size {len(datapipe)}")
+
+    mask_dir = "/data/constant_mask"
+    land_mask = torch.from_numpy(
```
Comment (dallasfoster) [inline examples/weather/pangu_weather/train_pangu_lite_era5.py:159]: Can the data type be configurable here? Either float32 or float16?

File: modulus/datapipes/climate/weatherbench.py
Comment (dallasfoster) [inline modulus/datapipes/climate/weatherbench.py:1]: Can we break this off as a separate MR.

File: modulus/models/fengwu/fengwu.py
Comment (dallasfoster) [inline modulus/models/fengwu/fengwu.py:1]: There is a lot of code in this file. Can we group together layers and subcomments into separate files? For example, let's put MLP, transformer blocks, attention blocks, and upsampling into models/layers or something.

File: modulus/models/fengwu/fengwu.py [hunk 9]
```diff
@@ -0,0 +1,1270 @@
...
+            output_resolution=resolution[0],
+            middle_resolution=resolution[1],
+            depth=2,
+            depth_middle=6,
+            num_heads=num_heads[:2],
+            window_size=window_size[1:],
+            drop_path=drop_path,
+        )
+
+    def forward(self, surface, z, r, u, v, t):
```
Comment (dallasfoster) [inline modulus/models/fengwu/fengwu.py:1222]: I think we should make it principle to have the forward call be strictly `def forward(self, x):` and the internal of the forward model separates the components as necessary.

File: modulus/models/fengwu/shift_window_mask.py
Comment (dallasfoster) [inline modulus/models/fengwu/shift_window_mask.py:1]: Consider putting these into modulus/models/utils folder.

File: modulus/models/fengwu/utils.py
Comment (dallasfoster) [inline modulus/models/fengwu/utils.py:1]: Again, consider putting these into modulus/models/utils

File: modulus/models/pangu/pangu.py [hunk 10]
```diff
@@ -0,0 +1,641 @@
...
+from torch import nn
+
+from ..meta import ModelMetaData
+from ..module import Module
+from .patch_embed import PatchEmbed2D, PatchEmbed3D, PatchRecovery2D, PatchRecovery3D
+from .shift_window_mask import get_shift_window_mask, window_partition, window_reverse
+from .utils import crop3d, get_earth_position_index, get_pad3d
+
+
+class UpSample(nn.Module):
```
Comment (dallasfoster) [inline modulus/models/pangu/pangu.py:32]: Is this class duplicated from Fengwu code? Let's consolidate all of the layers as much as possible.

File: modulus/models/pangu/pangu.py
Comment (dallasfoster) [inline modulus/models/pangu/pangu.py:1]: Similar comment to fengwu, let us separate out all of the layers into a separate folder and abstract them so that they can be shared by pangu, fengwu, etc...

File: modulus/models/swinvrnn/swinvrnn.py
Comment (dallasfoster) [inline modulus/models/swinvrnn/swinvrnn.py:1]: Similar comment about consolidation applies here.

**Review by dallasfoster — CHANGES_REQUESTED** [review]
Needs tidying of the layers and then we can rereview. Also consider merging the weatherbench datapipe separately, which will smooth the process.

**Comment by ivanauyeung** [comment]
Dear Dallas,

Thank you very much for the review. I will finish the following changes and make two new pull requests for your review. I will target to finish them within this week.

  1.  Tidy up the layers and make first pull request for model only.
  2.  Make second pull request for weatherbench datapipe.

Best Regards,
Ivan



From: Dallas Foster ***@***.***>
Date: Saturday, 18 May 2024 at 2:57 AM
To: NVIDIA/modulus ***@***.***>
Cc: Ivan Au Yeung ***@***.***>, Author ***@***.***>
Subject: Re: [NVIDIA/modulus] Added Pangu, Fengu, SwinRNN models (PR #341)

@dallasfoster requested changes on this pull request.

Needs tidying of the layers and then we can rereview. Also consider merging the weatherbench datapipe separately, which will smooth the process.

________________________________

In examples/weather/pangu_weather/train_pangu_era5.py<https://github.com/NVIDIA/modulus/pull/341#discussion_r1605417203>:

> +from torch.nn.parallel import DistributedDataParallel

+from omegaconf import DictConfig

+

+from modulus.models.pangu import Pangu

+from modulus.datapipes.climate import ERA5HDF5Datapipe

+from modulus.distributed import DistributedManager

+from modulus.utils import StaticCaptureTraining, StaticCaptureEvaluateNoGrad

+

+from modulus.launch.logging import LaunchLogger, PythonLogger, initialize_mlflow

+from modulus.launch.utils import load_checkpoint, save_checkpoint

+

+try:

+    from apex import optimizers

+except:

+    raise ImportError(

+        "FCN training requires apex package for optimizer."

Change the import error here.

________________________________

In examples/weather/pangu_weather/train_pangu_era5.py<https://github.com/NVIDIA/modulus/pull/341#discussion_r1605418283>:

> +    # Initialize loggers

+    # initialize_wandb(

+    #     project="Modulus-Launch-Dev",

+    #     entity="Modulus",

+    #     name="FourCastNet-Training",

+    #     group="FCN-DDP-Group",

+    # )

Let's remove this commented code.

________________________________

In examples/weather/pangu_weather/train_pangu_era5.py<https://github.com/NVIDIA/modulus/pull/341#discussion_r1605418602>:

> +        batch_size=1,

+        patch_size=(1, 1),

+        num_workers=8,

Can any of these be added to the config?

________________________________

In examples/weather/pangu_weather/train_pangu_era5.py<https://github.com/NVIDIA/modulus/pull/341#discussion_r1605419059>:

> +    no_channals_pangu = 4 + 5 * 13

+    datapipe = ERA5HDF5Datapipe(

+        data_dir="/data/train/",

+        stats_dir="/data/stats/",

+        channels=[i for i in range(no_channals_pangu)],

+        num_samples_per_year=cfg.num_samples_per_year_train,

+        batch_size=1,

+        patch_size=(1, 1),

+        num_workers=8,

+        device=dist.device,

+        process_rank=dist.rank,

+        world_size=dist.world_size,

+    )

+    logger.success(f"Loaded datapipe of size {len(datapipe)}")

+

+    mask_dir = "/data/constant_mask"

Can this be a configurable?

________________________________

In examples/weather/pangu_weather/train_pangu_era5.py<https://github.com/NVIDIA/modulus/pull/341#discussion_r1605419536>:

> +        img_size=(721, 1440),

+        patch_size=(2, 4, 4),

+        embed_dim=192,

+        num_heads=(6, 12, 12, 6),

+        window_size=(2, 6, 12),

Can any of this be configurable?

________________________________

In examples/weather/pangu_weather/train_pangu_lite_era5.py<https://github.com/NVIDIA/modulus/pull/341#discussion_r1605421654>:

> +    # Initialize loggers

+    # initialize_wandb(

+    #     project="Modulus-Launch-Dev",

+    #     entity="Modulus",

+    #     name="FourCastNet-Training",

+    #     group="FCN-DDP-Group",

+    # )

Remove commented code here.

________________________________

In examples/weather/pangu_weather/train_pangu_lite_era5.py<https://github.com/NVIDIA/modulus/pull/341#discussion_r1605421836>:

> +        batch_size=1,

+        patch_size=(1, 1),

+        num_workers=8,

Same comment, can these be configurable.

________________________________

In examples/weather/pangu_weather/train_pangu_lite_era5.py<https://github.com/NVIDIA/modulus/pull/341#discussion_r1605422169>:

> +    datapipe = ERA5HDF5Datapipe(

+        data_dir="/data/train/",

+        stats_dir="/data/stats/",

+        channels=[i for i in range(no_channals_pangu)],

+        num_samples_per_year=cfg.num_samples_per_year_train,

+        batch_size=1,

+        patch_size=(1, 1),

+        num_workers=8,

+        device=dist.device,

+        process_rank=dist.rank,

+        world_size=dist.world_size,

+    )

+    logger.success(f"Loaded datapipe of size {len(datapipe)}")

+

+    mask_dir = "/data/constant_mask"

+    land_mask = torch.from_numpy(

Can the data type be configurable here? Either float32 or float16?

________________________________

On modulus/datapipes/climate/weatherbench.py<https://github.com/NVIDIA/modulus/pull/341#discussion_r1605423398>:

Can we break this off as a separate MR.

________________________________

On modulus/models/fengwu/fengwu.py<https://github.com/NVIDIA/modulus/pull/341#discussion_r1605425877>:

There is a lot of code in this file. Can we group together layers and subcomments into separate files? For example, let's put MLP, transformer blocks, attention blocks, and upsampling into models/layers or something.

________________________________

In modulus/models/fengwu/fengwu.py<https://github.com/NVIDIA/modulus/pull/341#discussion_r1605426791>:

> +        )

+        self.decoder_t = DecoderLayer(

+            img_size=img_size,

+            patch_size=patch_size,

+            out_chans=pressure_level,

+            dim=embed_dim,

+            output_resolution=resolution[0],

+            middle_resolution=resolution[1],

+            depth=2,

+            depth_middle=6,

+            num_heads=num_heads[:2],

+            window_size=window_size[1:],

+            drop_path=drop_path,

+        )

+

+    def forward(self, surface, z, r, u, v, t):

I think we should make it principle to have the forward call be strictly def forward(self, x): and the internal of the forward model separates the components as necessary.

________________________________

On modulus/models/fengwu/shift_window_mask.py<https://github.com/NVIDIA/modulus/pull/341#discussion_r1605427430>:

Consider putting these into modulus/models/utils folder.

________________________________

On modulus/models/fengwu/utils.py<https://github.com/NVIDIA/modulus/pull/341#discussion_r1605428272>:

Again, consider putting these into modulus/models/utils

________________________________

In modulus/models/pangu/pangu.py<https://github.com/NVIDIA/modulus/pull/341#discussion_r1605429204>:

> +import math

+from dataclasses import dataclass

+

+import numpy as np

+import torch

+from timm.layers import DropPath, trunc_normal_

+from torch import nn

+

+from ..meta import ModelMetaData

+from ..module import Module

+from .patch_embed import PatchEmbed2D, PatchEmbed3D, PatchRecovery2D, PatchRecovery3D

+from .shift_window_mask import get_shift_window_mask, window_partition, window_reverse

+from .utils import crop3d, get_earth_position_index, get_pad3d

+

+

+class UpSample(nn.Module):

Is this class duplicated from Fengwu code? Let's consolidate all of the layers as much as possible.

________________________________

On modulus/models/pangu/pangu.py<https://github.com/NVIDIA/modulus/pull/341#discussion_r1605430074>:

Similar comment to fengwu, let us separate out all of the layers into a separate folder and abstract them so that they can be shared by pangu, fengwu, etc...

________________________________

On modulus/models/swinvrnn/swinvrnn.py<https://github.com/NVIDIA/modulus/pull/341#discussion_r1605430474>:

Similar comment about consolidation applies here.

File: examples/weather/pangu_weather/train_pangu_lite_era5.py [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+try:
+    from apex import optimizers
+except:
+    raise ImportError(
+        "FCN training requires apex package for optimizer."
+        + "See https://github.com/nvidia/apex for install details."
+    )
+
+
+def loss_func(x, y):
```
Comment (loliverhennigh) [inline examples/weather/pangu_weather/train_pangu_lite_era5.py:44]: I think you can replace this function with the pytorch l1 loss if I am not mistaken. https://pytorch.org/docs/stable/generated/torch.nn.L1Loss.html

File: examples/weather/pangu_weather/train_pangu_lite_era5.py [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+    pangu_model = Pangu(
+        img_size=(721, 1440),
+        patch_size=(2, 8, 8),
+        embed_dim=192,
+        num_heads=(6, 12, 12, 6),
+        window_size=(2, 6, 12),
+    ).to(dist.device)
+
+    if dist.rank == 0 and wandb.run is not None:
+        wandb.watch(
```
Comment (loliverhennigh) [inline examples/weather/pangu_weather/train_pangu_lite_era5.py:198]: I might be wrong but seems like we are not using wandb anymore or at least in this script. Seems like its using mlflow now?

File: examples/weather/pangu_weather/conf/config.yaml [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+hydra:
+  job:
+    chdir: True
+  run:
+    dir: ./outputs/
+
+wb_artifacts: False
```
Comment (loliverhennigh) [inline examples/weather/pangu_weather/conf/config.yaml:23]: I think this config is not used as its hard set for mlflow now.

File: examples/weather/pangu_weather/conf/config.yaml [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+use_mlflow: True
```
Comment (loliverhennigh) [inline examples/weather/pangu_weather/conf/config.yaml:24]: You can probably remove this from the configs and always set as true

**Review by loliverhennigh — COMMENTED** [review]
Few minor comments. I would echo @dallasfoster 's comments on better integrating the layers.

File: examples/weather/pangu_weather/train_pangu_era5.py (hunk 1 above)
Comment (ivanauyeung) [reply examples/weather/pangu_weather/train_pangu_era5.py]: Corrected the import error

File: examples/weather/pangu_weather/train_pangu_era5.py (hunk 2 above)
Comment (ivanauyeung) [reply examples/weather/pangu_weather/train_pangu_era5.py]: Removed the commented code

File: examples/weather/pangu_weather/train_pangu_era5.py (hunk 3 above)
Comment (ivanauyeung) [reply examples/weather/pangu_weather/train_pangu_era5.py]: Moved to conf/config.yaml

File: examples/weather/pangu_weather/train_pangu_era5.py (hunk 4 above)
Comment (ivanauyeung) [reply examples/weather/pangu_weather/train_pangu_era5.py]: Moved to conf/config.yaml

File: examples/weather/pangu_weather/train_pangu_era5.py (hunk 5 above)
Comment (ivanauyeung) [reply examples/weather/pangu_weather/train_pangu_era5.py]: Moved to conf/config.yaml

File: examples/weather/pangu_weather/train_pangu_lite_era5.py (hunk 6 above)
Comment (ivanauyeung) [reply examples/weather/pangu_weather/train_pangu_lite_era5.py]: Removed commented code

File: examples/weather/pangu_weather/train_pangu_lite_era5.py (hunk 7 above)
Comment (ivanauyeung) [reply examples/weather/pangu_weather/train_pangu_lite_era5.py]: Moved to conf/config.yaml

File: examples/weather/pangu_weather/train_pangu_lite_era5.py (hunk 8 above)
Comment (ivanauyeung) [reply examples/weather/pangu_weather/train_pangu_lite_era5.py:157]: Moved to conf/config.yaml

File: modulus/datapipes/climate/weatherbench.py
Comment (ivanauyeung) [reply modulus/datapipes/climate/weatherbench.py:1]: Removed weatherbench.py in this MR and will push another MR for it

File: modulus/models/fengwu/fengwu.py
Comment (ivanauyeung) [reply modulus/models/fengwu/fengwu.py:1]: Moved the layers into models/layers and consolidate with the layers for Pangu model

File: modulus/models/fengwu/fengwu.py (hunk 9 above)
Comment (ivanauyeung) [reply modulus/models/fengwu/fengwu.py]: Changed the forward to be one input. How about the output format?

File: modulus/models/fengwu/shift_window_mask.py
Comment (ivanauyeung) [reply modulus/models/fengwu/shift_window_mask.py:1]: Moved to models/utils and consolidate with the utils function for Pangu model

File: modulus/models/pangu/pangu.py (hunk 10 above)
Comment (ivanauyeung) [reply modulus/models/pangu/pangu.py]: Moved to models/layers

File: modulus/models/pangu/pangu.py
Comment (ivanauyeung) [reply modulus/models/pangu/pangu.py:1]: Moved to models/layers and consolidate with the layers for all model

File: modulus/models/swinvrnn/swinvrnn.py
Comment (ivanauyeung) [reply modulus/models/swinvrnn/swinvrnn.py:1]: Moved to models/layers and consolidate with the layers for all model

File: examples/weather/pangu_weather/conf/config.yaml (hunk 13 above)
Comment (ivanauyeung) [reply examples/weather/pangu_weather/conf/config.yaml]: Removed wandb related code

File: examples/weather/pangu_weather/conf/config.yaml (hunk 14 above)
Comment (ivanauyeung) [reply examples/weather/pangu_weather/conf/config.yaml]: Removed wandb related code

File: examples/weather/pangu_weather/train_pangu_era5.py [hunk 15]
```diff
@@ -0,0 +1,331 @@
...
+    dist = DistributedManager()
+
+    # Initialize loggers
+    # initialize_wandb(
+    #     project="Modulus-Launch-Dev",
+    #     entity="Modulus",
+    #     name="FourCastNet-Training",
+    #     group="FCN-DDP-Group",
+    # )
+    initialize_mlflow(
```
[the remark this answers is not in the document]
Comment (ivanauyeung) [reply examples/weather/pangu_weather/train_pangu_era5.py:121]: Moved the experiment name to config

File: examples/weather/pangu_weather/train_pangu_lite_era5.py (hunk 11 above)
Comment (ivanauyeung) [reply examples/weather/pangu_weather/train_pangu_lite_era5.py:43]: Replaced the loss to torch l1 loss

**Review by dallasfoster — COMMENTED** [review]
Changes look very good, just a few minor fixes and then we should be good.

File: examples/weather/pangu_weather/train_pangu_era5.py [hunk 16]
```diff
@@ -0,0 +1,322 @@
...
+        user_name="Modulus User",
+        mode="offline",
+    )
+    LaunchLogger.initialize(use_mlflow=True)  # Modulus launch logger
+    logger = PythonLogger("main")  # General python logger
+
+    no_channals_pangu = 4 + 5 * 13
+    datapipe = ERA5HDF5Datapipe(
+        data_dir="/data/train/",
+        stats_dir="/data/stats/",
```
[the remark this answers is not in the document]
Comment (ivanauyeung) [reply examples/weather/pangu_weather/train_pangu_era5.py]: Moved to conf/config.yaml

File: examples/weather/pangu_weather/train_pangu_era5.py [hunk 17]
```diff
@@ -0,0 +1,322 @@
...
+        experiment_desc=cfg.experiment_desc,
+        run_name="Pangu-trainng",
+        run_desc=cfg.experiment_desc,
+        user_name="Modulus User",
+        mode="offline",
+    )
+    LaunchLogger.initialize(use_mlflow=True)  # Modulus launch logger
+    logger = PythonLogger("main")  # General python logger
+
+    no_channals_pangu = 4 + 5 * 13
```
[the remark this answers is not in the document]
Comment (ivanauyeung) [reply examples/weather/pangu_weather/train_pangu_era5.py]: Fixed

File: examples/weather/pangu_weather/train_pangu_era5.py [hunk 18]
```diff
@@ -0,0 +1,322 @@
...
+    surface_mask = torch.stack([land_mask, soil_type, topography], dim=0).to(
+        dist.device
+    )
+    logger.success(f"Loaded suface constant mask from {mask_dir}")
+
+    if dist.rank == 0:
+        logger.file_logging()
+        validation_datapipe = ERA5HDF5Datapipe(
+            data_dir="/data/test/",
+            stats_dir="/data/stats/",
```
[the remark this answers is not in the document]
Comment (ivanauyeung) [reply examples/weather/pangu_weather/train_pangu_era5.py]: Moved to conf/config.yaml

File: modulus/models/layers/attention_layers.py [hunk 19]
```diff
@@ -0,0 +1,236 @@
...
+            window_size
+        )  # Wpl*Wlat*Wlon, Wpl*Wlat*Wlon
+        self.register_buffer("earth_position_index", earth_position_index)
+
+        self.qkv = nn.Linear(dim, dim * 3, bias=qkv_bias)
+        self.attn_drop = nn.Dropout(attn_drop)
+        self.proj = nn.Linear(dim, dim)
+        self.proj_drop = nn.Dropout(proj_drop)
+
+        trunc_normal_(self.earth_position_bias_table, std=0.02)
```
[the remark this answers is not in the document]
Comment (ivanauyeung) [reply modulus/models/layers/attention_layers.py]: Revised to "self.earth_position_bias_table = trunc_normal_(...)" to make use of "trunc_normal_"

File: modulus/models/layers/attention_layers.py [hunk 20]
```diff
@@ -0,0 +1,236 @@
...
+            window_size, ndim=2
+        )  # Wlat*Wlon, Wlat*Wlon
+        self.register_buffer("earth_position_index", earth_position_index)
+
+        self.qkv = nn.Linear(dim, dim * 3, bias=qkv_bias)
+        self.attn_drop = nn.Dropout(attn_drop)
+        self.proj = nn.Linear(dim, dim)
+        self.proj_drop = nn.Dropout(proj_drop)
+
+        trunc_normal_(self.earth_position_bias_table, std=0.02)
```
[the remark this answers is not in the document]
Comment (ivanauyeung) [reply modulus/models/layers/attention_layers.py]: Revised to "self.earth_position_bias_table = trunc_normal_(...)" to make use of "trunc_normal_"

**Review by dallasfoster — APPROVED** [review]

**Review by loliverhennigh — APPROVED** [review]
